File: webapp/src/lib/api/stats.ts
import { env } from '@/env'

/**
 * Homepage platform metrics. Hand-written to match the existing `lib/api/*.ts`
 * convention; the wire shape is the backend `PlatformStats` (snake_case keys).
 * Every scalar is nullable so the fallback pickers can decide live-vs-placeholder.
 */
export interface PlatformStats {
  catalog_total: number
  registries_count: number
  tier_distribution: Record<string, number>
  median_score: number | null
  p95_latency_ms: number | null
  avg_latency_ms: number | null
  rule_count: number
  agents_count: number
  github_stars: number | null
}

export async function fetchPlatformStats(): Promise<PlatformStats> {
  const res = await fetch(`${env.PUBLIC_API_URL}/api/v1/stats`, {
    headers: { Accept: 'application/json' },
  })
  if (!res.ok) throw new Error(`API ${res.status}`)
  return (await res.json()) as PlatformStats
}

File: services/api/app/schemas/stats.py
"""Hand-written `PlatformStats` response model for `GET /api/v1/stats`.

One payload backing every homepage metric. Snake_case keys (inherits
`OrmBaseModel`); the TS DTO falls out of `openapi.json` via the codegen
pipeline. Scalar metrics are nullable so the route can report "no data yet"
and let the frontend fall back to its launch placeholder.
"""

from __future__ import annotations

from pydantic import Field

from app.schemas.orm_base import OrmBaseModel


class PlatformStats(OrmBaseModel):
    catalog_total: int = Field(..., ge=0, description="Non-archived catalog item count.")
    registries_count: int = Field(..., ge=0, description="Distinct registries sourced from.")
    tier_distribution: dict[str, int] = Field(
        default_factory=dict, description="Item count bucketed by latest-scan tier."
    )
    median_score: int | None = Field(
        default=None, ge=0, le=100, description="P50 aggregate score over completed scans."
    )
    p95_latency_ms: int | None = Field(
        default=None, ge=0, description="P95 scan latency in ms over completed scans."
    )
    avg_latency_ms: int | None = Field(
        default=None, ge=0, description="Average scan latency in ms over completed scans."
    )
    rule_count: int = Field(..., ge=0, description="Active rubric rule count the engine loads.")
    agents_count: int = Field(..., ge=0, description="Supported agent platforms (static config).")
    github_stars: int | None = Field(
        default=None, ge=0, description="Repo star count (cached proxy); null when unavailable."
    )
